### PR TITLE
Implement a hierarchical rnn mnist example

### PR DESCRIPTION
**Pull request by iammrhelo**
Hierarchical RNNs can perform multiple levels of modeling. As in [ A Hierarchical Neural Autoencoder for Paragraphs and Documents ](https://web.stanford.edu/~jurafsky/pubs/P15-1107.pdf), a sentence is first encoded into the a sentence vector by the first layer of a rnn, then the second LSTM takes consecutive sentences vectors as input and encodes the sentence vector sequence into a paragraph vector

Hierarchical RNNs can be done elegantly in keras using TimeDistributed Wrappers and reshaping the Data into 4D. The hierarchical mnist model first encodes columns of pixels into a column vector, then encodes the column vector sequence into a image vector. Trained on MNIST, the training and validation accuracy reached 98% in 5 epochs.

screen shot 2016-08-13 at 02 23 05

### CODE REVIEW

File: examples/mnist_hierarchical_rnn.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+Y_train = np_utils.to_categorical(y_train, nb_classes)
+Y_test = np_utils.to_categorical(y_test, nb_classes)
+
+# Define model
+pixel_hidden = 128
+col_hidden   = 128 
+
+row, col, pixel = X_train.shape[1:]
+
+# 3D input
```
Comment (fchollet) [inline examples/mnist_hierarchical_rnn.py]: Technically it's 4D...

File: examples/mnist_hierarchical_rnn.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+"""
+An example of hierarchical RNN to predict MNIST digits
+Hierarchical model is implemented elegantly with the
```
Comment (fchollet) [inline examples/mnist_hierarchical_rnn.py]: No need to qualify the code as "elegant". The docstring needs more detailed explanations of what the model does, how, and why it is interesting (what you can apply it to, etc).

File: examples/mnist_hierarchical_rnn.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+pixel_hidden = 128
+col_hidden   = 128 
+
+row, col, pixel = X_train.shape[1:]
+
+# 3D input
+x = Input(shape=(row, col, pixel))
+
+# Encodes a col of pixels, uses TimeDistributed Wrapper 
+encoded_pixels = TimeDistributed( LSTM(output_dim=pixel_hidden) )(x)
```
Comment (fchollet) [inline examples/mnist_hierarchical_rnn.py]: No spaces around arguments.

File: examples/mnist_hierarchical_rnn.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+# Encodes columns of encoded pixels  
+encoded_columns = LSTM( col_hidden )(encoded_pixels)
```
Comment (fchollet) [inline examples/mnist_hierarchical_rnn.py]: Same.

File: examples/mnist_hierarchical_rnn.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+model = Model(input=[x], output=[prediction])
+
+model.compile(loss='categorical_crossentropy',
+              optimizer='rmsprop',
+              metrics=['accuracy'])
+
+model.fit(X_train, Y_train, batch_size=batch_size, nb_epoch=nb_epochs,
+          verbose=1, validation_data=(X_test, Y_test))
+
+scores = model.evaluate(X_test, Y_test, verbose=0)
```
Comment (fchollet) [inline examples/mnist_hierarchical_rnn.py]: Print something describing the results.

**Comment by iammrhelo** [comment]
@fchollet Recheck this, please~

File: examples/mnist_hierarchical_rnn.py (hunk 5 above)
Comment (iammrhelo) [inline examples/mnist_hierarchical_rnn.py]: Printed

**Comment by pifelbo** [comment]
A more common use case of a hierarchical RNN would be for sentences/words as mentioned in this example. However, simply modifying this example will not handle masking correctly as described in https://github.com/fchollet/keras/issues/4786.

I'd argue that until this issue gets fixed there should be a warning in this example as it's clearly describing this use case.
